common/src/main/java/io/netty/util/internal/DefaultPriorityQueue.java

```diff
@@ -0,0 +1,291 @@
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        if (!(o instanceof PriorityQueueNode)) {
+            return false;
+        }
+        PriorityQueueNode node = (PriorityQueueNode) o;
+        return contains(node, node.priorityQueueIndex(this));
+    }
+
+    @Override
+    public boolean containsTyped(T node) {
+        return contains(node, node.priorityQueueIndex(this));
+    }
+
+    @Override
+    public void clear() {
+        for (int i = 0; i < size; ++i) {
+            T node = queue[i];
+            if (node != null) {
+                node.priorityQueueIndex(this, INDEX_NOT_IN_QUEUE);
+                queue[i] = null;
+            }
+        }
+        size = 0;
+    }
+
+    @Override
+    public boolean offer(T e) {
+        if (e.priorityQueueIndex(this) != INDEX_NOT_IN_QUEUE) {
+            throw new IllegalArgumentException("e.priorityQueueIndex(): " + e.priorityQueueIndex(this) +
+                    " (expected: " + INDEX_NOT_IN_QUEUE + ") + e: " + e);
+        }
+
+        // Check that the array capacity is enough to hold values by doubling capacity.
+        if (size >= queue.length) {
+            // Use a policy which allows for a 0 initial capacity. Same policy as JDK's priority queue, double when
+            // "small", then grow by 50% when "large".
+            queue = Arrays.copyOf(queue, queue.length + ((queue.length < 64) ?
+                                                         (queue.length + 2) :
+                                                         (queue.length >>> 1)));
+        }
+
+        bubbleUp(size++, e);
+        return true;
+    }
+
+    @Override
+    public T poll() {
+        if (size == 0) {
+            return null;
+        }
+        T result = queue[0];
+        result.priorityQueueIndex(this, INDEX_NOT_IN_QUEUE);
+
+        T last = queue[--size];
+        queue[size] = null;
+        if (size != 0) { // Make sure we don't add the last element back.
+            bubbleDown(0, last);
+        }
+
+        return result;
+    }
+
+    @Override
+    public T peek() {
+        return (size == 0) ? null : queue[0];
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public boolean remove(Object o) {
+        final T node;
+        try {
+            node = (T) o;
+        } catch (ClassCastException e) {
+            return false;
+        }
+        return removeTyped(node);
+    }
+
+    @Override
+    public boolean removeTyped(T node) {
+        int i = node.priorityQueueIndex(this);
+        if (!contains(node, i)) {
+            return false;
+        }
+
+        node.priorityQueueIndex(this, INDEX_NOT_IN_QUEUE);
+        if (--size == 0 || size == i) {
+            // If there are no node left, or this is the last node in the array just remove and return.
+            queue[i] = null;
+            return true;
+        }
+
+        // Move the last element where node currently lives in the array.
+        T moved = queue[i] = queue[size];
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract the assignment out of this expression. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AAssignmentInSubExpressionCheck)